why-choice/index.qmd

```diff
@@ -0,0 +1,265 @@
+    [`datamodel_code_generator`](https://docs.pydantic.dev/latest/integrations/datamodel_code_generator/)
+    to convert the JSON schema of the standard to Pydantic models
+    automatically. We could then fetch and integrate updates to the
+    standard by regenerating the models.
+-   It would be easy to check package, resource and table metadata
+    separately.
+-   All errors are collected in a single round of checks.
+-   Error codes are detailed and easy to extract.
+-   Pydantic is widely used, extensively tested, and has good
+    documentation.
+:::
+
+::: column
+#### Drawbacks
+
+-   We would need to translate the Data Package standard into Pydantic
+    models to be able to use Pydantic for checking objects. However, we
+    could use `datamodel_code_generator` to automate this process (see
+    on the left).
+-   While integrating our existing metadata classes with Pydantic models
+    would lead to a single representation of the standard in Sprout, our
+    previous experience with generating classes from the schema has
+    shown that the resulting class system is more complex than what we
+    would like. If we choose Pydantic, a similar exercise of simplifying
+    the class system without losing compatibility with the standard is
+    likely to happen. This would lead to losing the ability to receive updates
+    to the standard automatically.
+-   As Pydantic models have a `schema` attribute, we would need to alias
+    the `schema` attribute of our resource metadata class internally to
+    avoid ambiguity.
+-   Currently, our metadata classes are allowed to represent partial
+    metadata objects (e.g. just one field specified). Adding checks
+    directly in these classes at instantiation time could lead to some
+    difficulties with instantiating partial metadata objects, which
+    might not meet the standard in that form.
+:::
+:::::
+
+### jsonschema
+
+[jsonschema](https://python-jsonschema.readthedocs.io/en/stable/) is the
+most widely used implementation of the [JSON Schema
+specification](https://json-schema.org/) in Python. Its main use case is
+checking that a JSON object conforms to a JSON schema.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Can check JSON metadata files against the JSON schema of the Data
+    Package standard directly.
+-   Matches the schema exactly, e.g., it disallows empty lists.
+-   Does not introduce a parallel set of classes representing the
+    standard.
+-   We can receive updates to the standard automatically by downloading
+    the new version of the schema.
+-   It is possible to collect all errors in a single round of checks.
+-   It is possible to filter errors both by source (which field did not
+    match) and by cause (why that field did not match).
+-   `jsonschema` is widely used, extensively tested, and has good
+    documentation.
+:::
+
+::: column
+#### Drawbacks
+
+-   Does not include optional recommendations in the textual description
+    of the standard.
```

Comment:
Is this also true for Pydantic?

Reply:
It's true in the sense that Pydantic models would include whatever we put in them, so by default they include nothing really.